src/libstore/binary-cache-store.cc

```diff
@@ -443,6 +443,25 @@ StorePath BinaryCacheStore::addTextToStore(const string & name, const string & s
     })->path;
 }
 
+std::optional<const Realisation> BinaryCacheStore::queryRealisation(const DrvOutput & id)
+{
+    auto outputInfoFilePath =
+        "/drvOutputs/" + std::string(id.drvPath.hashPart()) + "!" + id.outputName + ".doi";
+    auto rawOutputInfo = getFile(outputInfoFilePath);
+
+    if (rawOutputInfo) {
+        return { Realisation::parse(*rawOutputInfo, outputInfoFilePath) };
+    } else {
+        return std::nullopt;
+    }
+}
+
+void BinaryCacheStore::registerDrvOutput(const Realisation& info) {
+    auto filePath = "/drvOutputs/" + std::string(info.id.drvPath.hashPart()) +
+        "!" + info.id.outputName + ".doi";
```

Comment:
what do you all think about nesting further like this?
```suggestion
    auto filePath = "/drvOutputs/" + std::string(info.id.drvPath.hashPart()) +
        "/" + info.id.outputName + ".doi";
```
rather than using `!`

Reply:
I used `!` because that's the already standard delimiter for derivation outputs, so it makes sense to use the same on binary caches − I've actually pushed 391f753 to make this use `id.to_string()` directly which makes this more natural.

That being said, I don't have a very strong opinion, so a `/` could be good to

Reply:
We could even use a `/` for `id.to_string()`, but I like the fact that `!` gives a clear message of *this isn't a store path, don't try to use it as such*